agent/local/state.go

```diff
@@ -0,0 +1,963 @@
+		var err error
+		switch {
+		case c.Deleted:
+			err = l.deleteCheck(id)
+		case !c.InSync:
+			if c.DeferCheck != nil {
+				c.DeferCheck.Stop()
+				c.DeferCheck = nil
+			}
+			err = l.syncCheck(id)
+		default:
+			l.logger.Printf("[DEBUG] agent: Check '%s' in sync", id)
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	// Now sync the node level info if we need to, and didn't do any of
+	// the other sync operations.
+	if !l.nodeInfoInSync {
+		if err := l.syncNodeInfo(); err != nil {
+			return err
+		}
+	} else {
+		l.logger.Printf("[DEBUG] agent: Node info in sync")
+	}
+
+	return nil
+}
+
+// LoadMetadata loads node metadata fields from the agent config and
+// updates them on the local agent.
+func (l *State) LoadMetadata(data map[string]string) error {
+	l.Lock()
+	defer l.Unlock()
+
+	for k, v := range data {
+		l.metadata[k] = v
+	}
+	l.TriggerSyncChanges()
+	return nil
+}
+
+// UnloadMetadata resets the local metadata state
+func (l *State) UnloadMetadata() {
+	l.Lock()
+	defer l.Unlock()
+	l.metadata = make(map[string]string)
+}
+
+// Stats is used to get various debugging state from the sub-systems
+func (l *State) Stats() map[string]string {
+	l.RLock()
+	defer l.RUnlock()
+
+	services := 0
+	for _, s := range l.services {
+		if s.Deleted {
+			continue
+		}
+		services++
+	}
+
+	checks := 0
+	for _, c := range l.checks {
+		if c.Deleted {
+			continue
+		}
+		checks++
+	}
+
+	return map[string]string{
+		"services": strconv.Itoa(services),
+		"checks":   strconv.Itoa(checks),
+	}
+}
+
+// deleteService is used to delete a service from the server
+func (l *State) deleteService(id string) error {
+	if id == "" {
+		return fmt.Errorf("ServiceID missing")
+	}
+
+	req := structs.DeregisterRequest{
+		Datacenter:   l.config.Datacenter,
+		Node:         l.config.NodeName,
+		ServiceID:    id,
+		WriteRequest: structs.WriteRequest{Token: l.serviceToken(id)},
+	}
+	var out struct{}
+	err := l.Delegate.RPC("Catalog.Deregister", &req, &out)
+	if err == nil || strings.Contains(err.Error(), "Unknown service") {
+		delete(l.services, id)
+		l.logger.Printf("[INFO] agent: Deregistered service '%s'", id)
+		return nil
+	}
+	if acl.IsErrPermissionDenied(err) {
+		// todo(fs): why is the service in sync here?
+		l.services[id].InSync = true
+		l.logger.Printf("[WARN] agent: Service '%s' deregistration blocked by ACLs", id)
+		return nil
+	}
+	return err
+}
+
+// deleteCheck is used to delete a check from the server
+func (l *State) deleteCheck(id types.CheckID) error {
+	if id == "" {
+		return fmt.Errorf("CheckID missing")
+	}
+
+	req := structs.DeregisterRequest{
+		Datacenter:   l.config.Datacenter,
+		Node:         l.config.NodeName,
+		CheckID:      id,
+		WriteRequest: structs.WriteRequest{Token: l.checkToken(id)},
+	}
+	var out struct{}
+	err := l.Delegate.RPC("Catalog.Deregister", &req, &out)
+	if err == nil || strings.Contains(err.Error(), "Unknown check") {
+		// todo(fs): do we need to stop the deferCheck timer here?
```

Comment:
It reloads the check and skips out if the check is gone, so likely ok as-is, but stopping it is probably nice to do.

Reply:
stopping the timer now